targets/TARGET_NORDIC/TARGET_NRF5x/trng_api.c

```diff
@@ -0,0 +1,119 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "hal/trng_api.h"
+#include "hal/critical_section_api.h"
+
+#include "nrf_drv_rng.h"
+
+/* Keep track of instantiated FlashIAP objects. */
+static int nordic_trng_counter = 0;
+
+void trng_init(trng_t *obj)
+{
+    (void) obj;
+
+    /* Increment global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter++;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Initialize TRNG on first object only. */
+    if (counter == 1) {
+        nrf_drv_rng_init(NULL);
+    }
+}
+
+void trng_free(trng_t *obj)
+{
+    (void) obj;
+
+    /* Decrement global counter inside critical section. */
+    hal_critical_section_enter();
+    nordic_trng_counter--;
+    int counter = nordic_trng_counter;
+    hal_critical_section_exit();
+
+    /* Deinitialize TRNG when all objects have been freed. */
+    if (counter == 0) {
+        nrf_drv_rng_uninit();
+    }
+}
+
+/* Get random data from NRF5x TRNG peripheral.
+ *
+ * This implementation returns num of random bytes in range <1, length>.
+ * For parameters description see trng_api.h file.
+ */
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    (void) obj;
+
+    /* Query how many bytes are available. */
+    uint8_t bytes_available;
+    nrf_drv_rng_bytes_available(&bytes_available);
+
+    /* If no bytes are cached, block until at least 1 byte is available. */
+    if (bytes_available == 0) {
+        nrf_drv_rng_block_rand(output, 1);
+        *output_length = 1;
+    } else {
+
+        /* Get up to the requested number of bytes. */
+        if (bytes_available > length) {
+            bytes_available = length;
+        }
+
+        ret_code_t result = nrf_drv_rng_rand(output, bytes_available);
+
+        /* Set output length with available bytes. */
+        if (result == NRF_SUCCESS) {
+            *output_length = bytes_available;
+        } else {
+            *output_length = 0;
+        }
+    }
+
+    /* Set return value based on how many bytes was read. */
+    return (*output_length == 0) ? -1 : 0;
```

Comment:
In the corner case where `length == 0`, and where `trng_get_bytes` should be a no-op, this would return failure.

Reply:
Good point! Thanks!